ART regimen lines implementation

ssmusoke opened the pull request:
Reference https://app.asana.com/0/1131715315801874/1198214324308758 

Testing requires to use these PRS:

- UgandaEMR Reports https://github.com/METS-Programme/openmrs-module-ugandaemr-reports/pull/259

To test the PR:

1. Start the UgandaEMR server to install the concepts and program states, once successful restart to move all clients to the First Line Regimen 
2. Select a patient then move them to a new regimen status 
3. When editing the regimen change the type of switch cannot be edited due to side effects 

Updates - 29 July 2021
==================
1. When a patient is enrolled into care, must automatically enroll them into first line Regimen 
2. The Regimen Change form can only be completed when a patient has both an ART Summary and ART Clinical Encounters 

Pending:
1. Data validation rules for regimen switches not being recorded when the regimens change and regimen switch records not matching the regimen change visit dates 
2. ART Regimen Wizard shows the date enrolled and location name which are to be removed see Talk Post https://talk.openmrs.org/t/program-status-dashboard-widget-configuration/33939

ssmusoke committed:
Added ART_REGIMEN_CHANGE encounter

ssmusoke committed:
Added configuration and concepts for the program workflow states

ssmusoke committed:
Removed the ART Regimen Switch and Substitution sections which are moved to their own form

ssmusoke committed:
Regimen Change HTML Form

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

ssmusoke committed:
Add Regimen Change and Substitution concepts

ssmusoke committed:
Regimen Switch and Substitution form

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

 Conflicts:
	api/src/main/resources/metadata/Custom_Concepts.xml

ssmusoke committed:
Merged Regimen Change concepts with Covid 19 Concept Needs

ssmusoke committed:
Concepts from https://github.com/METS-Programme/openmrs-module-ugandaemr/pull/29 which includes those needed for this feature

ssmusoke committed:
Regimen change form - one missing feature which is to limit the regimens to those taken todate using the encounter date as a limiting

ssmusoke committed:
Changed new regimen to use the set not the list of concepts

ssmusoke committed:
Bump up to the latest version of legacyui which shows the program workflow states in the Cohort builder

ssmusoke committed:
Add liquibase changeset to setup all patients into the unknown state

ssmusoke committed:
Add the workflow states HTML form entry tag for the new line for switch and substitution

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

 Conflicts:
	api/src/main/java/org/openmrs/module/ugandaemr/api/deploy/bundle/CommonMetadataBundle.java
	api/src/main/java/org/openmrs/module/ugandaemr/metadata/core/EncounterTypes.java
	api/src/main/resources/liquibase.xml
	api/src/main/resources/metadata/Custom_Concepts.xml

ssmusoke committed:
The current line now displays the current regimen status of the patient

ssmusoke committed:
Added a widget to show the current regimen lines as well as changes over time

ssmusoke committed:
Added a test for validation rule for patients with Unknown regimen line

ssmusoke committed:
Update the liquibase changeset to run only when the program_workflow_state has been set rather then the program which probably exists

ssmusoke committed:
Change the sample data not to use the TB program

ssmusoke committed:
Remove unknown regimen line since all patients are defaulted to first line regimen

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

 Conflicts:
	api/src/main/resources/metadata/Custom_Concepts.xml

ssmusoke committed:
Add validation rules, and do not allow any regimen change or switch during editing since there are side effects

ssmusoke committed:
Change title of the dashboard widget to ART Regimen Lines

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

ssmusoke committed:
Clean up HTML Form Entry dependencies on upgrade to 4.x branch which contains the Velocity functions to limit obs and program workflow states by date

ssmusoke committed:
Remove the validation rule for patients on unknown regimen lines since as that line is not required anymore

slubwama commented:
@ssmusoke  resolved conflict

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

 Conflicts:
	api/src/main/resources/metadata/Custom_Concepts.xml

ssmusoke commented:
@slubwama conflict resolved

slubwama commented on api/src/test/java/org/openmrs/module/ugandaemr/dataintegrity/ARTRegimenLinesValidationRulesTest.java:
```diff
@@ -0,0 +1,34 @@
+
+	protected static final String UGANDAEMR_STANDARD_DATASET_XML = "org/openmrs/module/ugandaemr/include/standardTestDataset.xml";
+
+	ARTRegimenLinesValidationRules artRegimenLinesValidationRules;
```
@ssmusoke this throws an exception. Cant tell where it has to be imported from.

ssmusoke replied:
@slubwama I have deleted the test since the actual validation rule was removed from the implementation

ssmusoke committed:
Deleted validation rules test after the validation rule was removed

slubwama commented:
@ssmusoke  these validations  have got to go since the sections where the validations are being picked nolonger exist as observations on the page 

//Validation for Substitution within 1st line
            submitValStageART.push(dateValidator("99163_1", "99161", "99160", message5, message6, "less_than", false));
            submitValStageART.push(dateValidator("99163_2", "99161", "99160", message5, message6, "less_than", false));
            submitValStageART.push(dateValidator("99163_3", "99161", "99160", message5, message6, "less_than", false));
            submitValStageART.push(dateValidator("99163_4", "99161", "99160", message5, message6, "less_than", false));

            //Validation for Switch to 2nd
            submitValStageART.push(dateValidator("99164_1", "99161", "99160", message5, message6, "less_than", false));
            submitValStageART.push(dateValidator("99164_2", "99161", "99160", message5, message6, "less_than", false));
            submitValStageART.push(dateValidator("99164_3", "99161", "99160", message5, message6, "less_than", false));

            //Validation for Switch to 3rd
            submitValStageART.push(dateValidator("162991_1", "99161", "99160", message5, message6, "less_than", false));
            submitValStageART.push(dateValidator("162991_2", "99161", "99160", message5, message6, "less_than", false));
            submitValStageART.push(dateValidator("162991_3", "99161", "99160", message5, message6, "less_than", false));

slubwama commented:
<img width="414" alt="regimen" src="https://user-images.githubusercontent.com/3336745/127121315-92f61cfd-dd0b-4f73-841b-5cda14df0cf5.png">
Why is this still displaying function instead of data

ssmusoke commented:
> @ssmusoke these validations have got to go since the sections where the validations are being picked nolonger exist as observations on the page
> 
> //Validation for Substitution within 1st line
> submitValStageART.push(dateValidator("99163_1", "99161", "99160", message5, message6, "less_than", false));
> submitValStageART.push(dateValidator("99163_2", "99161", "99160", message5, message6, "less_than", false));
> submitValStageART.push(dateValidator("99163_3", "99161", "99160", message5, message6, "less_than", false));
> submitValStageART.push(dateValidator("99163_4", "99161", "99160", message5, message6, "less_than", false));
> 
> ```
>         //Validation for Switch to 2nd
>         submitValStageART.push(dateValidator("99164_1", "99161", "99160", message5, message6, "less_than", false));
>         submitValStageART.push(dateValidator("99164_2", "99161", "99160", message5, message6, "less_than", false));
>         submitValStageART.push(dateValidator("99164_3", "99161", "99160", message5, message6, "less_than", false));
> 
>         //Validation for Switch to 3rd
>         submitValStageART.push(dateValidator("162991_1", "99161", "99160", message5, message6, "less_than", false));
>         submitValStageART.push(dateValidator("162991_2", "99161", "99160", message5, message6, "less_than", false));
>         submitValStageART.push(dateValidator("162991_3", "99161", "99160", message5, message6, "less_than", false));
> ```

@slubwama Those changes are not in the RegimenChange file

ssmusoke commented:
> <img alt="regimen" width="414" src="https://user-images.githubusercontent.com/3336745/127121315-92f61cfd-dd0b-4f73-841b-5cda14df0cf5.png">
> 
> Why is this still displaying function instead of data

@slubwama Can you share a link of the page so that I can try to reproduce

slubwama commented:
> > <img alt="regimen" width="414" src="https://user-images.githubusercontent.com/3336745/127121315-92f61cfd-dd0b-4f73-841b-5cda14df0cf5.png">
> > Why is this still displaying function instead of data
> 
> @slubwama Can you share a link of the page so that I can try to reproduce

This is in the ART Regimen Switch and Substitution page. With a regimen as a baseline I thought that will be picked up and displayed. Only after using the ART Regimen Switch and Substitution page to make changes thats when changes are effected and the regimen will display 

<img width="619" alt="after switch " src="https://user-images.githubusercontent.com/3336745/127126011-5ddc0fc9-c21b-40d3-844e-eb381c3cb3e2.png">

slubwama commented:
@ssmusoke
1. When a new patient is enrolled on the summary page why is it that they are not enrolled into ART Program and also Assigned a regimen Line
2. On the Current Regimen Lookup ensure that the patient regimen is not only derived from 90315 Current regimen but also those that are from Transfer in regimen or Baseline regimen.

slubwama commented:
> > @ssmusoke these validations have got to go since the sections where the validations are being picked nolonger exist as observations on the page
> > //Validation for Substitution within 1st line
> > submitValStageART.push(dateValidator("99163_1", "99161", "99160", message5, message6, "less_than", false));
> > submitValStageART.push(dateValidator("99163_2", "99161", "99160", message5, message6, "less_than", false));
> > submitValStageART.push(dateValidator("99163_3", "99161", "99160", message5, message6, "less_than", false));
> > submitValStageART.push(dateValidator("99163_4", "99161", "99160", message5, message6, "less_than", false));
> > ```
> >         //Validation for Switch to 2nd
> >         submitValStageART.push(dateValidator("99164_1", "99161", "99160", message5, message6, "less_than", false));
> >         submitValStageART.push(dateValidator("99164_2", "99161", "99160", message5, message6, "less_than", false));
> >         submitValStageART.push(dateValidator("99164_3", "99161", "99160", message5, message6, "less_than", false));
> > 
> >         //Validation for Switch to 3rd
> >         submitValStageART.push(dateValidator("162991_1", "99161", "99160", message5, message6, "less_than", false));
> >         submitValStageART.push(dateValidator("162991_2", "99161", "99160", message5, message6, "less_than", false));
> >         submitValStageART.push(dateValidator("162991_3", "99161", "99160", message5, message6, "less_than", false));
> > ```
> 
> @slubwama Those changes are not in the RegimenChange file

@ssmusoke they are not but they will break the saving since the obs that  affect them have been removed in the PR

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

ssmusoke committed:
Add the ability to assign the patient to a regimen line (workflow state) in the HIV Program
Restrict the Regimen Change to happen only when a user has an ART Summary and Encounter page

ssmusoke commented:
@slubwama I have updated the PR see description, however my test and implementation for the Regimen assignment which does not seem to be working ....

Please take a look at https://github.com/METS-Programme/openmrs-module-ugandaemr/blob/15ba52a0c9321c2c44d26f06ca437cf49a68a7ba/omod/src/main/java/org/openmrs/module/ugandaemr/htmlformentry/ARTEnrollmentSubmissionAction.java#L48-#L56

and 

https://github.com/METS-Programme/openmrs-module-ugandaemr/blob/15ba52a0c9321c2c44d26f06ca437cf49a68a7ba/omod/src/test/java/org/openmrs/module/ugandaemr/htmlformentry/ARTEnrollmentSubmissionActionTest.java#L59-#L91

slubwama commented on api/src/test/resources/org/openmrs/module/ugandaemr/include/dsdmTestDataSet.xml:
```diff
@@ -46,18 +48,12 @@
     <concept_answer  concept_answer_id="7089"  concept_id="165143"  answer_concept="165141"  creator="1"  date_created="2018-10-15 13:02:40"  sort_weight="4"  uuid="0a2ac1f3-69ca-4b7d-99ba-0f55a512fa29"/>
     <concept_answer  concept_answer_id="7090"  concept_id="165143"  answer_concept="165142"  creator="1"  date_created="2018-10-15 13:02:40"  sort_weight="5"  uuid="cf6cea6f-c3e7-49f6-a447-15613f36504e"/>
 
-    <visit_type  visit_type_id="1"  name="Facility Visit"  description="Patient visits the clinic/hospital (as opposed to a home visit, or telephone contact)"  creator="1"  date_created="2013-08-02 03:39:43"  changed_by="1"  date_changed="2016-08-30 03:52:10"  retired="0"  uuid="7b0f5697-27e3-40c4-8bae-f4049abfb4ed"/>
```
@ssmusoke why did you delete visit type facility  from the data set.  You can only introduce in more data but not delete any. It might lead to a number of tests breaking

ssmusoke replied:
@slubwama Replaced, I had removed it when including the StandardDataset but there was too much duplication

ssmusoke committed:
Replace visit_type entry

slubwama commented on api/src/test/resources/org/openmrs/module/ugandaemr/include/dsdmTestDataSet.xml:
```diff
@@ -67,11 +65,19 @@
     <obs  obs_id="655624" status="FINAL" person_id="1393"  concept_id="165143"  encounter_id="31182"  obs_datetime="2015-02-10 00:00:00"  location_id="2"  value_coded="165138"  creator="1"  date_created="2019-01-18 14:26:45"  voided="0" uuid="b5d9efaf-c99e-4581-b834-3d5c3448d746"/>
     <obs  obs_id="357324" status="FINAL" person_id="1393"  concept_id="99161"  encounter_id="31181" obs_datetime="2010-03-12 00:00:00"    location_id="2"  value_datetime="2013-02-06 00:00:00" creator="1"  date_created="2015-02-18 17:23:55"  voided="0" uuid="8d131f53-66b6-4552-bbd9-c178e4470b6b"/>
 
+    <program program_id="1" concept_id="160631" creator="1" date_created="2021-07-29 12:08:19" changed_by="1" date_changed="2021-07-29 12:08:19" retired="0" name="HIV Program" description="Program where HIV patients are treated and monitored" uuid="18c6d4aa-0a36-11e7-8dbb-507b9dc4c741"/>
```
@ssmusoke this concept 160631 used in program can not be found. Introduce it in your data set.

ssmusoke replied:
@slubwama Apologies I had forgotten to bring back the changes, the error should be on the logic now

slubwama replied:
@ssmusoke did you solve this. I don't seem to see a commit that you solve this

ssmusoke replied:
@slubwama my apologies I had committed but not pushed

ssmusoke committed:
Added missing concept for HIV Program

ssmusoke committed:
Merge branch 'master' of https://github.com/METS-Programme/openmrs-module-ugandaemr into art_regimen_lines

ssmusoke committed:
Fix Javascript issues for removed fields

ssmusoke committed:
Change encounters to be loaded from the database so that the uuid values are consistent

ssmusoke commented:
> @ssmusoke these validations have got to go since the sections where the validations are being picked nolonger exist as observations on the page
> 
> //Validation for Substitution within 1st line
> submitValStageART.push(dateValidator("99163_1", "99161", "99160", message5, message6, "less_than", false));
> submitValStageART.push(dateValidator("99163_2", "99161", "99160", message5, message6, "less_than", false));
> submitValStageART.push(dateValidator("99163_3", "99161", "99160", message5, message6, "less_than", false));
> submitValStageART.push(dateValidator("99163_4", "99161", "99160", message5, message6, "less_than", false));
> 
> ```
>         //Validation for Switch to 2nd
>         submitValStageART.push(dateValidator("99164_1", "99161", "99160", message5, message6, "less_than", false));
>         submitValStageART.push(dateValidator("99164_2", "99161", "99160", message5, message6, "less_than", false));
>         submitValStageART.push(dateValidator("99164_3", "99161", "99160", message5, message6, "less_than", false));
> 
>         //Validation for Switch to 3rd
>         submitValStageART.push(dateValidator("162991_1", "99161", "99160", message5, message6, "less_than", false));
>         submitValStageART.push(dateValidator("162991_2", "99161", "99160", message5, message6, "less_than", false));
>         submitValStageART.push(dateValidator("162991_3", "99161", "99160", message5, message6, "less_than", false));
> ```

@slubwama Now I understand this, I have removed it from the ART Summary card

ssmusoke commented:
> <img alt="regimen" width="414" src="https://user-images.githubusercontent.com/3336745/127121315-92f61cfd-dd0b-4f73-841b-5cda14df0cf5.png">
> 
> Why is this still displaying function instead of data

@slubwama You need to update your server modules, since this PR uses the latest snapshot of HTML Form Entry 

1. Run `mvn clean install -U`
2. Run `mvn openmrs-sdk:deploy -Ddistro=org.openmrs.module:ugandaemr:3.2.0-SNAPSHOT` and you will be asked to select the server to upgrade

slubwama commented:
@ssmusoke this looks good to me after a long drive. Thanks
